### ISSUE TITLE
[constants] Brainstorm # 2: how we should deal with a not required, non-extensible enum of 1, with no client default

### ISSUE THREAD

**Issue by iscai-msft** [Azure/autorest.modelerfour#326]
What to do with a non-extensible enum of 1 that is used as a non-required parameter and doesn't have an `x-ms-default` specified in the swagger?

What I think: I think we should treat it similar to how we're treating a constant in the method signature. Meaning, I think we should have it as an optional parameter with a default value of the one enum.

There is also the question of whether we want to generate an enum in this case (@pakrym does .NET need it to be an enum)? For python it doesn't really matter, not sure about other languages. If we do want an enum in this case, m4 currently isn't making an enum for it, so this would be something that would need to be fixed on the m4 side.

**Comment by pakrym** [comment]
>@pakrym does .NET need it to be an enum

Depends what people intend when using a single value enum in this case. I though this is a pattern for constant values.
IIRC we don't generate the parameter and replace the parameter usage with the enum value.

@ShivangiReja can confirm.

**Comment by iscai-msft** [comment]
@pakrym In this case if you don't generate a parameter, the user wouldn't be able to nullify the value, which I think is a valid scenario for them, since the parameter is not required

**Comment by pakrym** [comment]
Then it shouldn't be defined as a single item enum and use the client default instead.

**Comment by daviwil** [comment]
Yeah, I think that this case is too ambiguous to know whether it's meant to be optional or not in terms of the service API.  I think this is either a problem of poorly specified Swagger or a lack of proper modelling capability in Modelerfour (which can be fixed).  Do we have a sense of how often this occurs in the specs repo?

**Comment by ShivangiReja** [comment]
>Depends what people intend when using a single value enum in this case. I though this is a pattern for constant values.
IIRC we don't generate the parameter and replace the parameter usage with the enum value.

>@ShivangiReja can confirm.

Correct @pakrym. We use it as a constant value.

**Comment by iscai-msft** [comment]
@daviwil I'll look into this the specs repo today and get back to you on the usage. I'm not sure if there is any, since this is part of the constants logic table that has every permutation of `x-ms-client-default`, required, extensible etc

bringing @lmazuel into the discussion of treating this like a constant

**Comment by iscai-msft** [comment]
@daviwil I'm not seeing too many instances in the current specs. Mainly if I see this scenario, it's in a response body, not in the request.

In compute I'm seeing a couple of cases, such as when [getting dedicated host groups](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/Microsoft.Compute/stable/2020-06-01/compute.json#L786). Seeing at most ~5 per API version in compute, so not too bad

Also seeing a couple in filesystem, namespaces etc, probably more around a total of 20-30. Overall, I think it could be feasible to change these swaggers, but it'd be a decent sized PR.

**Comment by jhendrixMSFT** [comment]
Blob storage also has this pattern, see ListContainersInclude and EncryptionAlgorithm parameters (query and header params respectively).

**Comment by pakrym** [comment]
This is still marked as a question. Is there something actionable as part of this issue of should it be closed?

**Comment by azabbasi** [comment]
We have this exact same issue where our swagger defines an optional header with only one enum value 
```
"if-none-match-star": {
      "name": "If-None-Match",
      "in": "header",
      "description": "Only perform the operation if the entity does not already exist.",
      "required": false,
      "type": "string",
      "enum": [
        "*"
      ],
      "x-ms-parameter-location": "method",
      "x-ms-parameter-grouping": {
        "postfix": "Options"
      }
    },
```

And the generator adds that header to the request by default even though the header is optional
